sdk/cosmos/azure_data_cosmos/src/cosmos_request.rs

```diff
@@ -0,0 +1,291 @@
+            resource_address: resource_id,
+            database_name: None,
+            collection_name: None,
+            document_name: None,
+            partition_key,
+            options,
+            is_name_based,
+            is_feed: false,
+            is_resource_name_parsed_from_uri: false,
+            use_gateway_mode: false,
+            use_status_code_for_failures: false,
+            use_status_code_for_403: false,
+            use_status_code_for_4041002: false,
+            use_status_code_for_429: false,
+            use_status_code_for_bad_request: false,
+            disable_archival_partition_not_found_retry: false,
+            disable_retry_with_policy: false,
+            force_name_cache_refresh: false,
+            force_partition_key_range_refresh: false,
+            force_collection_routing_map_refresh: false,
+            force_master_refresh: false,
+            last_collection_routing_map_hash_code: 0,
+            request_authorization_token_type: authorization_token_type,
+            partition_key_range_identity: None,
+            request_context: RequestContext::default(),
+            properties: HashMap::new(),
+            body,
+            query_string: None,
+            continuation: None,
+            entity_id: None,
+        }
+    }
+
+    /// Returns true if the operation does not modify server state and can be
+    /// treated as read-only for caching / retry heuristics.
+    pub fn is_read_only_request(&self) -> bool {
+        matches!(
+            self.operation_type,
+            OperationType::Read
+                | OperationType::Query
+                | OperationType::SqlQuery
+                | OperationType::QueryPlan
+                | OperationType::Head
+                | OperationType::HeadFeed
+        )
+    }
+
+    /// Maps the logical `OperationType` to its corresponding HTTP verb.
+    pub fn http_method(&self) -> Method {
+        match self.operation_type {
+            OperationType::Create
+            | OperationType::Upsert
+            | OperationType::Query
+            | OperationType::SqlQuery
+            | OperationType::Batch
+            | OperationType::QueryPlan => Method::Post,
+            OperationType::Delete => Method::Delete,
+            OperationType::Read => Method::Get,
+            OperationType::Replace => Method::Put,
+            OperationType::Patch => Method::Patch,
+            OperationType::Head | OperationType::HeadFeed => Method::Head,
+            _ => Method::Post,
+        }
+    }
+
+    /// Converts this `CosmosRequest` into a concrete `azure_core::http::Request`.
+    ///
+    /// Inserts partition key and (if present) item options as headers. For
+    /// write operations, sets JSON content type, upsert header (when applicable)
+    /// and attaches the body bytes. Panics if location routing information is
+    /// missing from `request_context`.
+    pub fn to_raw_request(&self) -> Request {
```

Comment:
Assuming the intent here is for the `CosmosRequest` to be destroyed, since this is probably the final step before sending the request along the wire, this would be more appropriate as an `into_raw_request` method that takes `self` by value. That would allow you to get rid of the clones, since you're taking ownership of those values.

Reply:
Refactored this.